File: test/csharp/SpecGate.CtscFixtures/Conformance/ComplexInputs/ComplexInputs.cs
using System.Globalization;
using SpecGate.Annotations;

namespace SpecGate.CtscFixtures.Conformance.ComplexInputs;

[SpecEvent("EnumMemberInput")]
public sealed class EnumMemberInput
{
    [SpecEvent("name")]
    public string Name { get; set; } = string.Empty;

    [SpecEvent("value")]
    public string Value { get; set; } = string.Empty;
}

[SpecEvent("Point")]
public sealed class Point
{
    [SpecEvent("x")]
    public int X { get; set; }

    [SpecEvent("y")]
    public int Y { get; set; }
}

[SpecEvent("AppConfig")]
public sealed class AppConfig
{
    [SpecEvent("name")]
    public string Name { get; set; } = string.Empty;

    [SpecEvent("max_retries")]
    public int MaxRetries { get; set; }

    [SpecEvent("verbose")]
    public bool Verbose { get; set; }
}

[SpecEvent("Shape")]
public abstract class Shape;

[SpecEvent("Circle")]
public sealed class Circle : Shape
{
    [SpecEvent("radius")]
    public int Radius { get; set; }
}

[SpecEvent("Rectangle")]
public sealed class Rectangle : Shape
{
    [SpecEvent("width")]
    public int Width { get; set; }

    [SpecEvent("height")]
    public int Height { get; set; }
}

[SpecEvent("Point")]
public sealed class ShapePoint : Shape;

[SpecEvent("Address")]
public sealed class Address
{
    [SpecEvent("street")]
    public string Street { get; set; } = string.Empty;

    [SpecEvent("city")]
    public string City { get; set; } = string.Empty;
}

[SpecEvent("Person")]
public sealed class Person
{
    [SpecEvent("name")]
    public string Name { get; set; } = string.Empty;

    [SpecEvent("age")]
    public int Age { get; set; }

    [SpecEvent("address")]
    public Address Address { get; set; } = new();
}

public static class ComplexInputs
{
    [SpecOperation("create_enum_type", Spec = "fixture.complex_inputs")]
    public static string CreateEnumType(
        [SpecInput("name")] string name,
        [SpecInput("members")] List<EnumMemberInput> members)
    {
        _ = MemberCount(members);
        _ = FirstMember(members);
        return name;
    }

    [SpecOperation("member_count", Spec = "fixture.complex_inputs")]
    public static int MemberCount([SpecInput("members")] List<EnumMemberInput> members) =>
        members.Count;

    [SpecOperation("first_member", Spec = "fixture.complex_inputs")]
    public static string FirstMember([SpecInput("members")] List<EnumMemberInput> members) =>
        members.Count == 0 ? string.Empty : members[0].Name;

    [SpecOperation("sum_points", Spec = "fixture.complex_inputs")]
    public static Point SumPoints([SpecInput("points")] List<Point> points) =>
        new()
        {
            X = points.Sum(point => point.X),
            Y = points.Sum(point => point.Y),
        };

    [SpecOperation("describe_config", Spec = "fixture.complex_inputs")]
    public static string DescribeConfig([SpecInput("config")] AppConfig config) => config.Name;

    [SpecOperation("area_of_shape", Spec = "fixture.complex_inputs")]
    public static int AreaOfShape([SpecInput("shape")] Shape shape) =>
        shape switch
        {
            Circle circle => (int)(Math.PI * circle.Radius * circle.Radius),
            Rectangle rectangle => rectangle.Width * rectangle.Height,
            _ => 0,
        };

    [SpecOperation("classify", Spec = "fixture.complex_inputs")]
    public static Shape Classify([SpecInput("sides")] int sides) =>
        sides switch
        {
            4 => new Rectangle { Width = 3, Height = 4 },
            1 => new ShapePoint(),
            _ => new Circle { Radius = 5 },
        };

    [SpecOperation("get_points_on_line", Spec = "fixture.complex_inputs")]
    public static List<Point> GetPointsOnLine([SpecInput("count")] int count)
    {
        List<Point> points = [];
        for (int value = 0; value < count; value++)
        {
            points.Add(new Point { X = value, Y = value });
        }

        return points;
    }

    [SpecOperation("lookup", Spec = "fixture.complex_inputs")]
    public static int Lookup(
        [SpecInput("table")] SortedDictionary<string, int> table,
        [SpecInput("key")] string key) =>
        table.GetValueOrDefault(key);

    [SpecOperation("invert_map", Spec = "fixture.complex_inputs")]
    public static SortedDictionary<string, string> InvertMap(
        [SpecInput("table")] SortedDictionary<string, int> table)
    {
        SortedDictionary<string, string> inverted = [];
        foreach (KeyValuePair<string, int> entry in table)
        {
            inverted[entry.Value.ToString(CultureInfo.InvariantCulture)] = entry.Key;
        }

        return inverted;
    }

    [SpecOperation("greet_optional", Spec = "fixture.complex_inputs")]
    public static string GreetOptional([SpecInput("name")] Option<string> name) =>
        name.HasValue ? $"Hello, {name.Value}!" : "Hello, stranger!";

    [SpecOperation("find_point", Spec = "fixture.complex_inputs")]
    public static Option<Point> FindPoint(
        [SpecInput("points")] List<Point> points,
        [SpecInput("target_x")] int targetX)
    {
        Point? point = points.Find(point => point.X == targetX);
        return point is null ? Option<Point>.None() : Option<Point>.Some(point);
    }

    [SpecOperation("find_shape", Spec = "fixture.complex_inputs")]
    public static Option<Shape> FindShape([SpecInput("sides")] int sides) =>
        sides switch
        {
            1 => Option<Shape>.Some(new Circle { Radius = 5 }),
            0 => Option<Shape>.Some(new ShapePoint()),
            _ => Option<Shape>.None(),
        };

    [SpecOperation("describe_person", Spec = "fixture.complex_inputs")]
    public static string DescribePerson([SpecInput("person")] Person person) =>
        $"{person.Name}, age {person.Age}";

    [SpecOperation("create_person", Spec = "fixture.complex_inputs")]
    public static Person CreatePerson(
        [SpecInput("name")] string name,
        [SpecInput("age")] int age,
        [SpecInput("street")] string street,
        [SpecInput("city")] string city) =>
        new()
        {
            Name = name,
            Age = age,
            Address = new Address { Street = street, City = city },
        };
}

File: test/rust/crates/specgate-ctsc-fixtures/src/conformance/complex_inputs/complex_inputs.rs
use specgate::{SpecEvent, spec_operation};
use std::collections::BTreeMap;

#[derive(Debug, Clone, PartialEq, Eq, SpecEvent)]
#[spec_component("fixture.complex_inputs")]
pub struct EnumMemberInput {
    #[spec_event]
    pub name: String,
    #[spec_event]
    pub value: String,
}

#[derive(Debug, Clone, PartialEq, Eq, SpecEvent)]
#[spec_component("fixture.complex_inputs")]
pub struct Point {
    #[spec_event]
    pub x: i32,
    #[spec_event]
    pub y: i32,
}

#[derive(Debug, Clone, PartialEq, Eq, SpecEvent)]
#[spec_component("fixture.complex_inputs")]
pub struct AppConfig {
    #[spec_event]
    pub name: String,
    #[spec_event]
    pub max_retries: i32,
    #[spec_event]
    pub verbose: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, SpecEvent)]
#[spec_component("fixture.complex_inputs")]
pub enum Shape {
    Circle { radius: i32 },
    Rectangle { width: i32, height: i32 },
    Point,
}

#[derive(Debug, Clone, PartialEq, Eq, SpecEvent)]
#[spec_component("fixture.complex_inputs")]
pub struct Address {
    #[spec_event]
    pub street: String,
    #[spec_event]
    pub city: String,
}

#[derive(Debug, Clone, PartialEq, Eq, SpecEvent)]
#[spec_component("fixture.complex_inputs")]
pub struct Person {
    #[spec_event]
    pub name: String,
    #[spec_event]
    pub age: i32,
    #[spec_event]
    pub address: Address,
}

#[spec_operation("create_enum_type", spec = "fixture.complex_inputs")]
pub fn create_enum_type(name: &str, members: Vec<EnumMemberInput>) -> String {
    member_count(&members);
    first_member(&members);
    name.to_string()
}

#[spec_operation("member_count", spec = "fixture.complex_inputs")]
pub fn member_count(members: &[EnumMemberInput]) -> i32 {
    i32::try_from(members.len()).unwrap_or(i32::MAX)
}

#[spec_operation("first_member", spec = "fixture.complex_inputs")]
pub fn first_member(members: &[EnumMemberInput]) -> String {
    members
        .first()
        .map_or_else(String::new, |member| member.name.clone())
}

#[spec_operation("sum_points", spec = "fixture.complex_inputs")]
pub fn sum_points(points: Vec<Point>) -> Point {
    Point {
        x: points.iter().map(|point| point.x).sum(),
        y: points.iter().map(|point| point.y).sum(),
    }
}

#[spec_operation("describe_config", spec = "fixture.complex_inputs")]
pub fn describe_config(config: AppConfig) -> String {
    config.name
}

#[spec_operation("area_of_shape", spec = "fixture.complex_inputs")]
pub fn area_of_shape(shape: Shape) -> i32 {
    match shape {
        Shape::Circle { radius } => (std::f64::consts::PI * f64::from(radius * radius)) as i32,
        Shape::Rectangle { width, height } => width * height,
        Shape::Point => 0,
    }
}

#[spec_operation("classify", spec = "fixture.complex_inputs")]
pub fn classify(sides: i32) -> Shape {
    match sides {
        4 => Shape::Rectangle {
            width: 3,
            height: 4,
        },
        1 => Shape::Point,
        _ => Shape::Circle { radius: 5 },
    }
}

#[spec_operation("get_points_on_line", spec = "fixture.complex_inputs")]
pub fn get_points_on_line(count: i32) -> Vec<Point> {
    (0..count)
        .map(|value| Point { x: value, y: value })
        .collect()
}

#[spec_operation("lookup", spec = "fixture.complex_inputs")]
pub fn lookup(table: BTreeMap<String, i32>, key: &str) -> i32 {
    table.get(key).copied().unwrap_or(0)
}

#[spec_operation("invert_map", spec = "fixture.complex_inputs")]
pub fn invert_map(table: BTreeMap<String, i32>) -> BTreeMap<String, String> {
    table
        .into_iter()
        .map(|(key, value)| (value.to_string(), key))
        .collect()
}

#[spec_operation("greet_optional", spec = "fixture.complex_inputs")]
pub fn greet_optional(name: Option<String>) -> String {
    name.map_or_else(
        || "Hello, stranger!".to_string(),
        |name| format!("Hello, {name}!"),
    )
}

#[spec_operation("find_point", spec = "fixture.complex_inputs")]
pub fn find_point(points: Vec<Point>, target_x: i32) -> Option<Point> {
    points.into_iter().find(|point| point.x == target_x)
}

#[spec_operation("find_shape", spec = "fixture.complex_inputs")]
pub fn find_shape(sides: i32) -> Option<Shape> {
    match sides {
        1 => Some(Shape::Circle { radius: 5 }),
        0 => Some(Shape::Point),
        _ => None,
    }
}

#[spec_operation("describe_person", spec = "fixture.complex_inputs")]
pub fn describe_person(person: Person) -> String {
    format!("{}, age {}", person.name, person.age)
}

#[spec_operation("create_person", spec = "fixture.complex_inputs")]
pub fn create_person(name: &str, age: i32, street: &str, city: &str) -> Person {
    Person {
        name: name.to_string(),
        age,
        address: Address {
            street: street.to_string(),
            city: city.to_string(),
        },
    }
}

#[test]
fn structured_inputs_and_outputs_use_real_values() {
    assert_eq!(
        create_enum_type(
            "Color",
            vec![
                EnumMemberInput {
                    name: "Red".to_string(),
                    value: "1".to_string(),
                },
                EnumMemberInput {
                    name: "Blue".to_string(),
                    value: "2".to_string(),
                },
            ],
        ),
        "Color"
    );
    assert_eq!(
        sum_points(vec![Point { x: 1, y: 2 }, Point { x: 3, y: 4 }]),
        Point { x: 4, y: 6 }
    );
    assert_eq!(
        describe_config(AppConfig {
            name: "demo".to_string(),
            max_retries: 3,
            verbose: true,
        }),
        "demo"
    );
    assert_eq!(
        area_of_shape(Shape::Rectangle {
            width: 3,
            height: 4
        }),
        12
    );
    assert_eq!(
        classify(4),
        Shape::Rectangle {
            width: 3,
            height: 4
        }
    );
    assert_eq!(
        get_points_on_line(3),
        vec![
            Point { x: 0, y: 0 },
            Point { x: 1, y: 1 },
            Point { x: 2, y: 2 }
        ]
    );
    let table = BTreeMap::from([("alpha".to_string(), 7), ("beta".to_string(), 9)]);
    assert_eq!(lookup(table.clone(), "beta"), 9);
    assert_eq!(
        invert_map(table),
        BTreeMap::from([
            ("7".to_string(), "alpha".to_string()),
            ("9".to_string(), "beta".to_string())
        ])
    );
    assert_eq!(
        greet_optional(Some("Ada".to_string())),
        "Hello, Ada!".to_string()
    );
    assert_eq!(greet_optional(None), "Hello, stranger!".to_string());
    assert_eq!(
        find_point(vec![Point { x: 2, y: 8 }, Point { x: 4, y: 16 }], 4),
        Some(Point { x: 4, y: 16 })
    );
    assert_eq!(find_shape(0), Some(Shape::Point));
    let person = create_person("Grace", 37, "2 Main", "Arlington");
    assert_eq!(person.address.city, "Arlington");
    assert_eq!(describe_person(person), "Grace, age 37");
}
